File: BOJ/[1260] DFS와 BFS.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
#define MAX 10001
using namespace std;

vector<int> graph[MAX];
bool visited[MAX];
int N, M, V;

void reset()
{
    fill(visited, &visited[10000], false);
    printf("\n");
}

void dfs(int v)
{
    visited[v] = true;
    printf("%d ", v);
    for (int i = 0; i < graph[v].size(); i++)
    {
        if (!visited[graph[v][i]])
        {
            visited[graph[v][i]] = true;
            dfs(graph[v][i]);
        }
    }
}

void bfs()
{
    queue<int> Q;
    Q.push(V);
    visited[V] = true;

    int v;
    while (!Q.empty())
    {
        v = Q.front();
        Q.pop();
        printf("%d ", v);

        for (int i = 0; i < graph[v].size(); i++)
        {
            if (!visited[graph[v][i]])
            {
                visited[graph[v][i]] = true;
                Q.push(graph[v][i]);
            }
        }
    }
}

int main()
{
    scanf("%d%d%d", &N, &M, &V);

    int v1, v2;
    for (int i = 0; i < M; i++)
    {
        scanf("%d%d", &v1, &v2);
        graph[v1].push_back(v2);
        graph[v2].push_back(v1);
    }

    for (int i = 0; i < N; ++i)
        sort(graph[i].begin(), graph[i].end());

    dfs(V);
    reset();
    bfs();
}

File: BOJ/[1260] DFS와 BFS.py
import sys
from collections import deque, defaultdict

N, M, V = map(int, sys.stdin.readline().split())
graph = defaultdict(list)
for _ in range(M):
    v1, v2 = map(int, sys.stdin.readline().split())
    graph[v1].append(v2)
    graph[v2].append(v1)

# DFS
for i in range(1, N + 1):
    graph[i].sort(reverse=True)

result_dfs = []
is_visited = [False] * (N + 1)
stack = deque([V])
while stack:
    popped = stack.pop()
    if is_visited[popped]:
        continue

    is_visited[popped] = True
    result_dfs.append(popped)
    for node in graph[popped]:
        if not is_visited[node]:
            stack.append(node)

# BFS
for i in range(1, N + 1):
    graph[i].sort()

result_bfs = []
is_visited = [False] * (N + 1)
queue = deque([V])
is_visited[V] = True
while queue:
    popped = queue.popleft()
    result_bfs.append(popped)
    for node in graph[popped]:
        if not is_visited[node]:
            is_visited[node] = True
            queue.append(node)

for number in result_dfs:
    print(number, end=" ")
print()
for number in result_bfs:
    print(number, end=" ")
print()
